### PR TITLE
add string:equalsIgnoreCaseAscii langlib method

### LINKED ISSUE #21238: Add string:equalsIgnoreCaseAscii
**Description:**
Please refer https://github.com/ballerina-platform/ballerina-spec/issues/428

**Steps to reproduce:**
ballerina 1.1.0

### PR DESCRIPTION
**Pull request by dulvinw**
Fixes #21238

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    static {
+        decoder = Charset.forName("US-ASCII").newDecoder();
+    }
+
+    public static boolean equalsIgnoreCaseAscii(Strand strand, String str1, String str2) {
+        if (str1.length() != str2.length()) {
+            return false;
+        }
+
+        for (int i=0; i<str1.length(); i++) {
```
Comment (anupama-pathirage) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]:
```suggestion
        for (int i = 0; i <str1.length(); i++) {
```

Comment (anupama-pathirage) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Isn't this captured as a check style issue during the build?

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            String charAtIfromOne = Character.toString(str1.charAt(i));
+            String charAtIfromTwo = Character.toString(str2.charAt(i));
+            if (!(isPureAscii(charAtIfromOne) && isPureAscii(charAtIfromTwo))) {
+                if (!charAtIfromOne.equals(charAtIfromTwo)) {
+                    return false;
+                }
+            }
+            if (!charAtIfromOne.equalsIgnoreCase(charAtIfromTwo)) {
```
Comment (MaryamZi) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: We reach this point even for `!(isPureAscii(charAtIfromOne) && !(isPureAscii(charAtIfromTwo))` right? We shouldn't be doing `equalsIgnoreCase` in that case?

Comment (MaryamZi) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Can't we simplify this logic by handling the flow in the following order?

1. if both are pure ASCII, return `equalsIgnoreCase`
2. else return `equals`

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            return false;
+        }
+
+        for (int i = 0; i <str1.length(); i++) {
+            String charFromOne = Character.toString(str1.charAt(i));
+            String charFromTwo = Character.toString(str2.charAt(i));
+            if (isPureAscii(charFromOne) && isPureAscii(charFromTwo)) {
+                if (!charFromOne.equalsIgnoreCase(charFromTwo)) {
+                    return false;
+                }
```
Comment (MaryamZi) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]:
```suggestion
                return charFromOne.equalsIgnoreCase(charFromTwo);
```

Comment (MaryamZi) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: As discussed offline, shall we change these checks to something like

```java
            boolean charFromOnePureAscii = isPureAscii(charFromOne);
            boolean charFromTwoPureAscii = isPureAscii(charFromTwo);

            if (charFromOnePureAscii && charFromTwoPureAscii) {
                return charFromOne.equalsIgnoreCase(charFromTwo);
            } else if (!charFromOnePureAscii && !charFromTwoPureAscii) {
                return charFromOne.equals(charFromTwo);
            }
            return false;  
```

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    private static boolean isPureAscii(String  str) {
+        byte byteArray[] = str.getBytes();
+        try {
+            decoder.decode(ByteBuffer.wrap(byteArray)).toString();
+        } catch (CharacterCodingException e) {
+            return false;
+        }
+
```
Comment (MaryamZi) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]:
```suggestion
```

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java (hunk 3 above)
Comment (MaryamZi) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Sorry, missed the loop. This is OK.

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java (hunk 2 above)
Comment (dulvinw) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Fixed

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                }
+            } else if (!charFromOne.equals(charFromTwo)) {
+                return false;
+            }
+
+        }
+        return true;
+    }
+
+    private static boolean isPureAscii(String  str) {
```
Comment (dulajdilshan) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]:
```suggestion
    private static boolean isPureAscii(String str) {
```

File: langlib/lang.string/src/main/ballerina/src/lang.string/string.bal [hunk 6]
```diff
@@ -154,3 +154,11 @@ public function toCodePointInts(string str) returns int[] = external;
 # + return - a string with a character for each code point in `codePoints`; or an error
 # if any member of `codePoints` is not a valid code point
 public function fromCodePointInts(int[] codePoints) returns string|error = external;
+
+# Compare two given strings ignoring the case. In the comparison, any non ascii character
```
Comment (pubudu91) [inline langlib/lang.string/src/main/ballerina/src/lang.string/string.bal]:
```suggestion
# Compare two given strings ignoring the case. In the comparison, any non ASCII character
```

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                }
+            } else if (!charFromOne.equals(charFromTwo)) {
+                return false;
+            }
+
+        }
+        return true;
+    }
+
+    private static boolean isPureAscii(String str) {
```
Comment (pubudu91) [inline langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Do we need to make it this complicated? Isn't checking whether the char value is <127 enough?

File: langlib/lang.string/src/main/ballerina/src/lang.string/string.bal (hunk 6 above)
Comment (MaryamZi) [reply langlib/lang.string/src/main/ballerina/src/lang.string/string.bal]: @dulvinw, seems like this documentation is different from what is in the spec. Shall we change to use the same?

https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/lib/string.bal#L127

Comment (dulvinw) [reply langlib/lang.string/src/main/ballerina/src/lang.string/string.bal]: Fixed

File: langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java (hunk 7 above)
Comment (dulvinw) [reply langlib/lang.string/src/main/java/org/ballerinalang/langlib/string/EqualsIgnoreCaseAscii.java]: Will add an issue to improve performance

**Pull request merged**